eval/mt_bench/components.py

```diff
@@ -44,40 +45,34 @@ def launch_vllm_server_background(model_path: str, retries: int = 60, delay: int
         
         raise RuntimeError(f"Failed to start vLLM server at {server_url} after {retries} retries.")
 
-    def stop_vllm_server(server_pid: int):
-        import subprocess
-        import time
-        import os
-        import signal
+    # This seems like excessive effort to stop the vllm process, but merely saving & killing the pid doesn't work
+    # Also, the base image does not include `pkill` cmd, so can't pkill -f vllm.entrypoints.openai.api_server either
+    def stop_vllm_server_by_name():
+        import psutil
+
+        for process in psutil.process_iter(attrs=["pid", "name", "cmdline"]):
+            cmdline = process.info.get("cmdline")
+            if cmdline and "vllm.entrypoints.openai.api_server" in cmdline:
+                print(f"Found vLLM server process with PID: {process.info['pid']}, terminating...")
+                try:
+                    process.terminate()  # Try graceful termination
+                    process.wait(timeout=5)  # Wait a bit for it to terminate
+                    if process.is_running():
+                        print(f"Forcefully killing vLLM server process with PID: {process.info['pid']}")
+                        process.kill()  # Force kill if it's still running
+                    print(f"Successfully stopped vLLM server with PID: {process.info['pid']}")
+                except psutil.NoSuchProcess:
+                    print(f"Process with PID {process.info['pid']} no longer exists.")
+                except psutil.AccessDenied:
+                    print(f"Access denied when trying to terminate process with PID {process.info['pid']}.")
+                except Exception as e:
+                    print(f"Failed to terminate process with PID {process.info['pid']}. Error: {e}")
```

Comment:
It was not working in the first place because vLLM does not handle signals properly. It expects to be run from the CLI and thus stopped via SIGINT.

If you use a process group in the instantiation above you can get ride of the current pid search logic and simply do:

```python
process_group_id = os.getpgid(process.pid)
process.send_signal(signal.SIGINT)
try:
    print("Waiting for vLLM server to shut down gracefully")
    process.wait(timeout)
except subprocess.TimeoutExpired:
    print(
        f"Sending SIGKILL to vLLM server since timeout ({timeout}s) expired"
    )
    process.kill()

# Attempt to cleanup any remaining child processes
# Make sure process_group is legit (> 1) before trying
if process_group_id and process_group_id > 1:
    try:
        os.killpg(process_group_id, signal.SIGKILL)
        print("Sent SIGKILL to vLLM process group")
    except ProcessLookupError:
        print("Nothing left to clean up with the vLLM process group")
else:
    print("vLLM process group id not found")
```

Reply:
Thanks! Please add in a follow-up, I need to be done with this PR 🤣